feature/s3/transfermanager/api_op_PutObject.go

```diff
@@ -0,0 +1,1077 @@
+	}
+	return bytes.NewReader(part[0:n]), n, cleanup, err
+}
+
+func readFillBuf(r io.Reader, b []byte) (offset int, err error) {
+	for offset < len(b) && err == nil {
+		var n int
+		n, err = r.Read(b[offset:])
+		offset += n
+	}
+	return offset, err
+}
+
+type multiUploader struct {
+	*uploader
+	wg       sync.WaitGroup
+	m        sync.Mutex
+	err      error
+	uploadID *string
+	parts    completedParts
+}
+
+type ulChunk struct {
+	buf     io.Reader
+	partNum *int32
+	cleanup func()
+}
+
+type completedParts []CompletedPart
+
+func (cp completedParts) Len() int {
+	return len(cp)
+}
+
+func (cp completedParts) Less(i, j int) bool {
+	return aws.ToInt32(cp[i].PartNumber) < aws.ToInt32(cp[j].PartNumber)
+}
+
+func (cp completedParts) Swap(i, j int) {
+	cp[i], cp[j] = cp[j], cp[i]
+}
+
+// upload will perform a multipart upload using the firstBuf buffer containing
+// the first chunk of data.
+func (u *multiUploader) upload(ctx context.Context, firstBuf io.Reader, cleanup func(), clientOptions ...func(*s3.Options)) (*PutObjectOutput, error) {
+	params := u.uploader.in.mapCreateMultipartUploadInput()
+
+	// Create a multipart
+	resp, err := u.uploader.options.S3.CreateMultipartUpload(ctx, params, clientOptions...)
+	if err != nil {
+		cleanup()
+		return nil, err
+	}
+	u.uploadID = resp.UploadId
+
+	ch := make(chan ulChunk, u.options.Concurrency)
+	for i := 0; i < u.options.Concurrency; i++ {
+		// launch workers
+		u.wg.Add(1)
+		go u.readChunk(ctx, ch, clientOptions...)
+	}
+
+	var partNum int32 = 1
+	ch <- ulChunk{buf: firstBuf, partNum: aws.Int32(partNum), cleanup: cleanup}
+	for u.geterr() == nil && err == nil {
+		partNum++
+		var (
+			data         io.Reader
+			nextChunkLen int
+			ok           bool
+		)
+		data, nextChunkLen, cleanup, err = u.nextReader(ctx)
+		ok, err = u.shouldContinue(partNum, nextChunkLen, err)
+		if !ok {
+			cleanup()
+			if err != nil {
+				u.seterr(err)
+			}
+			break
+		}
+
+		ch <- ulChunk{buf: data, partNum: aws.Int32(partNum), cleanup: cleanup}
+	}
+
+	// close the channel, wait for workers and complete upload
+	close(ch)
+	u.wg.Wait()
+	completeOut := u.complete(ctx, clientOptions...)
+
+	if err := u.geterr(); err != nil {
+		return nil, &multipartUploadError{
+			err:      err,
+			uploadID: *u.uploadID,
+		}
+	}
+
+	var out PutObjectOutput
+	out.mapFromCompleteMultipartUploadOutput(completeOut, aws.ToString(params.Bucket), aws.ToString(u.uploadID), u.parts)
+	return &out, nil
+}
+
+func (u *multiUploader) shouldContinue(part int32, nextChunkLen int, err error) (bool, error) {
+	if err != nil && err != io.EOF {
+		return false, fmt.Errorf("read multipart upload data failed, %w", err)
+	}
+
+	if nextChunkLen == 0 {
+		// No need to upload empty part, if file was empty to start
+		// with empty single part would of been created and never
+		// started multipart upload.
+		return false, nil
+	}
+
+	// This upload exceeded maximum number of supported parts, error now.
+	if part > DefaultMaxUploadParts {
```

Comment:
This should be `MaxUploadParts` then instead of just `DefaultMax...`. But same question as previous, why do we want to limit the number of parts used?

Reply:
It appears there is in fact a 10k part limit for MPUs, I didn't catch this at first.

We're going to need to discuss offline how we want to handle this. The error case we're hitting here is unacceptable (failing a 10,000 part MPU only after we realize there are 10k parts).